src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Selectors/GenericSecurityTokenAuthenticator.cs

```diff
@@ -21,7 +31,46 @@ protected override bool CanValidateTokenCore(SecurityToken token)
         protected override ReadOnlyCollection<IAuthorizationPolicy> ValidateTokenCore(SecurityToken token)
         {
             var genericToken = (GenericSecurityToken)token;
-            return CoreWCF.Security.SecurityUtils.CreatePrincipalNameAuthorizationPolicies(genericToken.Name);
+            string principalName = genericToken.Name;
+            if (principalName == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(principalName));
+            }
+
+            Claim identityClaim;
+            Claim primaryPrincipal;
+            if (principalName.Contains("@") || principalName.Contains(@"\"))
+            {
+                identityClaim = new Claim(ClaimTypes.Upn, principalName, Rights.Identity);
+                primaryPrincipal = Claim.CreateUpnClaim(principalName);
+            }
+            else
+            {
+                identityClaim = new Claim(ClaimTypes.Spn, principalName, Rights.Identity);
+                primaryPrincipal = Claim.CreateSpnClaim(principalName);
+            }
+
+            List<Claim> claims = new List<Claim>(2)
+            {
+                identityClaim,
+                primaryPrincipal
+            };
+
+
+            if (_ldapSettings != null)
+            {
+                List<Claim> allCaims = LdapAdapter.RetrieveClaimsAsync(_ldapSettings, genericToken.GenericIdentity.Name).Result;
+                // if this is made async, many other API changes has to happen. COnsidering this is one of the scenario, ok to take the hit ?
+                foreach (Claim claim in allCaims)
```

Comment:
This is one of the improvements we should make to the api's. Can you open an issue to make SecurityTokenAuthenticator async?

Reply:
https://github.com/CoreWCF/CoreWCF/issues/375